src/v/storage/mvlog/versioned_log.cc

```diff
@@ -0,0 +1,171 @@
+    co_await active_seg_->segment_file->flush();
+    auto ro_seg = std::make_unique<readonly_segment>(std::move(active_seg_));
+    segs_.emplace_back(std::move(ro_seg));
+}
+
+ss::future<> versioned_log::create_unlocked(model::offset base) {
+    vassert(
+      !active_segment_lock_.ready(),
+      "create_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ == nullptr, "Expected no active segment");
+    auto new_path = fmt::format(
+      "{}/{}.log", ntp_cfg_.base_directory(), next_segment_id_());
+    vlog(log.info, "Creating new segment file {}", new_path);
+    auto f = co_await file_mgr_.create_file(std::filesystem::path{new_path});
+    auto active_seg = std::make_unique<active_segment>(
+      std::move(f), base, next_segment_id_, compute_max_segment_size());
+    active_seg_ = std::move(active_seg);
+    ++next_segment_id_;
+}
+
+ss::future<> versioned_log::apply_segment_ms() {
+    auto lock = active_segment_lock_.try_get_units();
+    if (!lock.has_value()) {
+        lock = co_await active_segment_lock_.get_units();
+    }
+    if (active_seg_ == nullptr) {
+        co_return;
+    }
+    const auto target_roll_deadline = compute_roll_deadline();
+    if (!target_roll_deadline.has_value()) {
+        co_return;
+    }
+    const auto now = ss::lowres_clock::now();
+    if (now >= target_roll_deadline.value()) {
+        vlog(
+          log.debug,
+          "Starting to roll {}, now vs deadline: {} vs {}",
+          ss::sstring(active_seg_->segment_file->filepath()),
+          now.time_since_epoch(),
+          target_roll_deadline->time_since_epoch());
+        co_await roll_unlocked();
+    }
+}
+
+ss::future<> versioned_log::close() {
+    // TODO(awong): handle concurrency. Should also prevent further appends,
+    // readers, etc.
+    if (active_seg_) {
+        co_await active_seg_->segment_file->close();
+    }
+    for (auto& seg : segs_) {
+        co_await seg->segment_file->close();
+    }
+}
+
+ss::future<> versioned_log::append(model::record_batch b) {
+    auto lock = active_segment_lock_.try_get_units();
+    if (!lock.has_value()) {
+        lock = co_await active_segment_lock_.get_units();
+    }
+    if (active_seg_ != nullptr) {
+        const auto cur_size = active_seg_->segment_file->size();
+        const auto target_max_size = active_seg_->target_max_size;
+        if (cur_size >= target_max_size) {
+            vlog(
+              log.debug,
+              "Starting to roll {}, cur bytes vs target bytes: {} vs {}",
+              active_seg_->segment_file->filepath().c_str(),
+              cur_size,
+              target_max_size);
+            // The active segment needs to be rolled.
+            co_await roll_unlocked();
+        }
+    }
+    if (active_seg_ == nullptr) {
+        co_await create_unlocked(b.base_offset());
+    }
+    vassert(active_seg_ != nullptr, "Expected active segment");
+    auto next = model::next_offset(b.last_offset());
```

Comment:
What do you think about also asserting `b.first_offset() == active_seg_->next_offset`?

Reply:
Yeah I think long term it probably makes sense to return some error code here. Added an assertion and a TODO